src/client/mod.rs

```diff
@@ -478,23 +480,113 @@ header! {
+        let decoded: Value = parse_vault_response(res)?;
+        decoded.get("sealed").and_then(|v| v.as_bool())
+               .map(|sealed| if sealed { VaultStatus::Sealed } else { VaultStatus::Unsealed })
+               .ok_or(Error::Vault("expected seal status in response".to_owned()))
+    }
+
+    /// Make an API request to progress through unsealing a vault. Return if already unsealed.
+    #[allow(unused_results)]
+    fn unseal<U, T>(host: U, keys: &[T]) -> Result<()>
+        where U: TryInto<Url, Err = Error>, T: Clone + Into<String>
+    {
+        let host = host.try_into()?;
+        let client = Client::new()?;
+        let res = handle_hyper_response(client.get(host.join("/v1/sys/seal-status")?).send())?;
+        let decoded: Value = parse_vault_response(res)?;
+        let mut sealed = decoded.get("sealed").and_then(|v| v.as_bool())
+                                .ok_or(Error::Vault("expected seal status in response".to_owned()))?;
+        let mut key_iter = keys.iter();
+        if sealed {
+            let total = decoded.get("t").and_then(|v| v.as_u64())
+                               .ok_or(Error::Vault("expected total number of keys in response".to_owned()))?;
+            if keys.len() < total as usize {
+                return Err(Error::Vault(format!("expected at least {} keys for unsealing Vault", total)))
+            }
+        }
+
+        while sealed {
+            let key: String = key_iter.next().unwrap().clone().into();
+            let payload = json!({"key": key});
+            let res = handle_hyper_response(client.put(host.join("/v1/sys/unseal")?)
+                                                  .body(payload.to_string().as_str())
+                                                  .send())?;
+            let decoded: Value = parse_vault_response(res)?;
+            sealed = decoded.get("sealed").and_then(|v| v.as_bool())
+                            .ok_or(Error::Vault("expected seal status in response".to_owned()))?;
+        }
+
+        Ok(())
+    }
+
+    /// Construct a `VaultClient` from an existing vault token
+    pub fn new<U, T, S>(host: U, token: T, unseal_keys: &[S]) -> Result<VaultClient<TokenData>>
+        where U: TryInto<Url, Err = Error> + Clone, T: Clone + Into<String>, S: Clone + Into<String>
+    {
+        VaultClient::unseal(host.clone(), unseal_keys)?;
+        let host = host.try_into()?;
+        let client = Client::new()?;
+        let res = handle_hyper_response(client.get(host.join("/v1/auth/token/lookup-self")?)
+                                              .header(XVaultToken(token.clone().into()))
+                                              .send())?;
         let decoded: VaultResponse<TokenData> = parse_vault_response(res)?;
         Ok(VaultClient {
             host: host,
-            token: token,
+            token: token.into(),
             client: client,
+            keys: unseal_keys.iter().map(|s| s.clone().into()).collect(),
```

Comment:
Right, but the token (also need not be a root token) is actually used throughout the workflows, e.g. when we set `X-Vault-Token` header so having it available as part of the client makes sense because we need it frequently.

Here we're never referencing `self.keys`. The `unseal` function is really a free-standing function (which I believe is correct): `fn unseal<U, T>(host: U, keys: &[T]) -> Result<()>`.

I also think the `new` constructor is a bit odd now because we're saying that the token is required but now also the unseal keys. You won't be able to get a token to begin with if the vault is already sealed.

I'm not sure coupling the unsealing with making a new vault client is the right approach.